src/auth/User.cc

```diff
@@ -47,6 +47,21 @@ Auth::User::credentials(CredentialState newCreds)
     credentials_state = newCreds;
 }
 
+void
+Auth::User::noteHelperTtl(const char *ttlNote)
+{
+    const int64_t ttl = (ttlNote ? strtoll(ttlNote, nullptr, 10) : -1);
+    updateExpiration(ttl);
```

Comment:
updateExpiration() interprets negavite TTLs as zero. Can we simplify and do the same here instead of using `-1` for the "default" value?

Reply:
updateExpiration() uses a scheme-specific default, OR zero if no scheme-specific override of that method is defined (yet).


We could redesign how Squid does defaults.  If we make credentialsttl a shared member that all schemes configure to their default, this code could use that member instead of -1.   That would leaves us with negative not being anything special and updaetExpiration() could be allowed to set expiry times in the past.

Reply:
> updateExpiration() uses a scheme-specific default, OR zero if no scheme-specific override of that method is defined (yet).

The virtual Auth::User::updateExpiration() method is documented (and implemented) to treat negative TTL values as zero. That documentation is an API contract -- any override in a derived class will have to do exactly the same. Thus, calling updateExpiration() with a `0` value is equivalent to calling it with `-1`. Thus, we should be calling it with zero because zero has a much simpler semantics (expires at the end of this second) than `-1` (expires after some default time).

Now, it is possible that you do not like the outcome of the above changes. I am not saying that the outcome is "correct"! I am only saying that it is logical based on the current PR code. If you do not like the outcome, then something else has to change in the PR code. If nothing relevant changes, then the above logic will lead to the outcome I outlined.

> We could redesign how Squid does defaults...

Let's not broaden this specific change request. If you would like to redesign something else in this PR, that is your call, of course, but let's keep this simple change request focused on one simple change. If future/anticipated changes are going to obsolete this change request, please say so (and then revisit this change request when those changes happen).

Reply:
I am thinking the fix for this change request is to accept that the design requirement (to retain special meaning on -1 from old design) that led to this situation was a bad one and can be dropped.

So yes. I propose a different mechanism (described paragraph 2 of https://github.com/squid-cache/squid/pull/774#discussion_r579868080) which allows ttl to have any integer value, no special case.

Reply:
There are probably many ways to address the problem identified in this change request. I have not studied the issue enough to recommend a specific solution. If you think that the best solution is to eliminate the special meaning of `-1` TTL by providing the actual default value via some suitable class member (e.g., a virtual defaultCredentialsTtl() method), then let's try that. Please do not forget to adjust Auth::User::updateExpiration() documentation accordingly.

Reply:
Done in aa2e693